initial commit

songhappy opened the pull request:
#### Context
What is the purpose of this PR? Is it to
- [x] fix a bug

Please link to any issues this PR addresses.
https://jira.devtools.intel.com/browse/IPB-2799
#### Changelog
What are the changes made in this PR?
*
Add xpu_support in verify_bf16_support, the only change needed is to verify if both torch.xpu.is_available() and torch.xpu.is_bf16_supported() are true.
#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] manually run any new or modified recipes with sufficient proof of correctness

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [x] I did not change any public API

ebsmothers commented:
Hi @songhappy thanks for the PR. Is the purpose here to enable training on XPU in torchtune? If so, can you share more context in the summary (e.g. is the bf16 support check the only change that needs to be made)?

songhappy commented:
@ebsmothers thank you very much for reviewing, I just rebased with new main branch, addressed in the summary that the only change needed is to verify bf16 support, it is very minor change.

SalmanMohammadi commented:
Hey @songhappy. Thanks for this PR. Could you please try run a recipe with `xpu` and `bf16` and attach some outputs to this PR to make sure everything works okay?

songhappy commented:
> Hey @songhappy. Thanks for this PR. Could you please try run a recipe with `xpu` and `bf16` and attach some outputs to this PR to make sure everything works okay?

Here is the command:
`tune run full_finetune_single_device --config recipes/configs/llama3_2/3B_full_single_device_my.yaml`
output looks like:
```
  0%|                                                                           | 0/13000 [00:00<?, ?it/s]
1|1|Loss: 2.958725690841675:   0%|                                    | 1/13000 [00:02<8:27:35,  2.34s/it]
1|2|Loss: 1.4943794012069702:   0%|                                   | 2/13000 [00:03<6:36:57,  1.83s/it]
```

songhappy commented on torchtune/utils/_device.py:
```diff
@@ -186,14 +188,15 @@ class DeviceSupport(Enum):
-    3. `communication_backend` (str): Specifies the backend used for communication on this device (e.g., "gloo", "nccl", "hccl").
+    1. `device_type` (str): The type of device (e.g., "cpu", "cuda", "npu", "xpu").
+    2. `device_name` (str): A user-friendly name for the device (e.g., "CPU", "GPU", "NPU", "XPU").
+    3. `communication_backend` (str): Specifies the backend used for communication on this device (e.g., "gloo", "nccl", "hccl", "ccl").
```
done

SalmanMohammadi commented:
@songhappy could you please run our pre-commit hooks to lint your code? There's instructions here https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md#coding-style

songhappy commented:
> @songhappy could you please run our pre-commit hooks to lint your code? There's instructions here https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md#coding-style

many thanks, I have addressed all the comments, all ready to merger, please approve :)

SalmanMohammadi commented on torchtune/utils/_device.py:
```diff
@@ -186,14 +188,16 @@ class DeviceSupport(Enum):
     This is a simple enum for compute devices,
     This currently only supports CPU, CUDA, NPU.
```
```suggestion
    This currently only supports CPU, CUDA, NPU, and XPU.
```

SalmanMohammadi reviewed (approved):
One last thing and we're good to go. Stamping to unblock - thanks so much for adding this!
